File: labs-opensource/M03B-context-engineering/starter/diagnose.js
/**
 * M03B Lab: diagnose.js (COMPLETE — runs against YOUR context_budget.js)
 * ========================================================================
 * Demonstrates the poisoned-transcript effect and the fix.
 * Run: node diagnose.js
 */

import { readFileSync } from "node:fs";
import OpenAI from "openai";
import { ContextBudget, summarizeHistory } from "./context_budget.js";

const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });

async function askModel(budget) {
  const { system, messages } = budget.buildMessages();
  const t0 = Date.now();
  try {
    const response = await client.chat.completions.create({
      model: "mistral",
      messages: [{ role: "system", content: system }, ...messages],
      max_tokens: 300,
    });
    return [
      response.choices[0].message.content ?? "",
      {
        input_tokens: response.usage?.prompt_tokens ?? 0,
        output_tokens: response.usage?.completion_tokens ?? 0,
        latency_s: (Date.now() - t0) / 1000,
      },
    ];
  } catch (e) {
    return [`[error: ${e.message}]`, { input_tokens: 0, output_tokens: 0, latency_s: 0 }];
  }
}

const fixture = JSON.parse(readFileSync(new URL("./poisoned_transcript.json", import.meta.url), "utf-8"));

const baseArgs = {
  model: "mistral",
  systemPrompt: fixture.system_prompt,
  toolDefinitions: fixture.tool_definitions,
  history: fixture.history,
  currentUserMessage: fixture.current_user_message,
};

// --- Run 1: rotted context ---
const rotted = new ContextBudget(baseArgs);
console.log("=== Token Breakdown (rotted) ===");
for (const [layer, tok] of Object.entries(rotted.account())) {
  console.log(`  ${layer.padEnd(15)}: ${tok.toLocaleString()} tokens`);
}
console.log(
  `  ${"TOTAL".padEnd(15)}: ${rotted.total().toLocaleString()} / ${rotted.maxTokens.toLocaleString()}  strategy=${rotted.strategy()}`
);

console.log("\n>>> Run 1: ROTTED context (no fix)");
const [answerA, usageA] = await askModel(rotted);
console.log(`Tokens: ${usageA.input_tokens} in, ${usageA.output_tokens} out  (${usageA.latency_s.toFixed(2)}s)`);
console.log(`Answer: ${answerA}\n`);

// --- Run 2: after checkpoint ---
const fixed = new ContextBudget({ ...baseArgs, history: [...fixture.history] });
fixed.history = await summarizeHistory(fixed.history, 4);
console.log(">>> Run 2: COMPRESSED context (after checkpoint)");
console.log(`  Total after compression: ${fixed.total().toLocaleString()} tokens  strategy=${fixed.strategy()}`);
const [answerB, usageB] = await askModel(fixed);
console.log(`Tokens: ${usageB.input_tokens} in, ${usageB.output_tokens} out  (${usageB.latency_s.toFixed(2)}s)`);
console.log(`Answer: ${answerB}\n`);

console.log("=".repeat(60));
console.log(`Token delta:   ${usageA.input_tokens - usageB.input_tokens} input tokens`);
console.log(`Latency delta: ${(usageA.latency_s - usageB.latency_s).toFixed(2)}s`);
console.log("\nSuccess check: Run 2 must still cite ORD-88421 and November 3rd.");

File: labs-opensource/M03B-context-engineering/starter/diagnose.py
#!/usr/bin/env python3
"""
M03B Lab: diagnose.py (COMPLETE — runs against YOUR context_budget.py)
========================================================================
Demonstrates the poisoned-transcript effect and the fix.
Run: python diagnose.py
"""

import json
import time

from openai import OpenAI
from context_budget import ContextBudget, summarize_history

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")


def ask_model(budget: ContextBudget) -> tuple[str, dict]:
    system, messages = budget.build_messages()
    t0 = time.time()
    try:
        response = client.chat.completions.create(
            model="mistral",
            messages=[{"role": "system", "content": system}] + messages,
            max_tokens=300,
        )
        answer = response.choices[0].message.content or ""
        usage = {
            "input_tokens": response.usage.prompt_tokens,
            "output_tokens": response.usage.completion_tokens,
            "latency_s": round(time.time() - t0, 2),
        }
    except Exception as e:
        answer, usage = f"[error: {e}]", {"input_tokens": 0, "output_tokens": 0, "latency_s": 0}
    return answer, usage


def main():
    with open("poisoned_transcript.json", encoding="utf-8") as f:
        fixture = json.load(f)

    base_args = dict(
        model="mistral",
        system_prompt=fixture["system_prompt"],
        tool_definitions=fixture["tool_definitions"],
        history=fixture["history"],
        current_user_message=fixture["current_user_message"],
    )

    # --- Run 1: rotted context ---
    rotted = ContextBudget(**base_args)
    breakdown = rotted.account()
    print("=== Token Breakdown (rotted) ===")
    for layer, tok in breakdown.items():
        print(f"  {layer:15s}: {tok:,} tokens")
    print(f"  {'TOTAL':15s}: {rotted.total():,} / {rotted.max_tokens:,}  strategy={rotted.strategy()!r}")

    print("\n>>> Run 1: ROTTED context (no fix)")
    answer_a, usage_a = ask_model(rotted)
    print(f"Tokens: {usage_a['input_tokens']} in, {usage_a['output_tokens']} out  ({usage_a['latency_s']}s)")
    print(f"Answer: {answer_a}\n")

    # --- Run 2: after checkpoint ---
    fixed = ContextBudget(**base_args)
    fixed.history = summarize_history(fixed.history, keep_recent=4)
    print(">>> Run 2: COMPRESSED context (after checkpoint)")
    print(f"  Total after compression: {fixed.total():,} tokens  strategy={fixed.strategy()!r}")
    answer_b, usage_b = ask_model(fixed)
    print(f"Tokens: {usage_b['input_tokens']} in, {usage_b['output_tokens']} out  ({usage_b['latency_s']}s)")
    print(f"Answer: {answer_b}\n")

    print("=" * 60)
    print(f"Token delta:   {usage_a['input_tokens'] - usage_b['input_tokens']:+,} input tokens")
    print(f"Latency delta: {usage_a['latency_s'] - usage_b['latency_s']:+.2f}s")
    print("\nSuccess check: Run 2 must still cite ORD-88421 and November 3rd.")


if __name__ == "__main__":
    main()
